controllers/user.go

```diff
@@ -253,6 +253,45 @@ func (c *ApiController) GetUser() {
 	c.ResponseOk(maskedUser)
 }
 
+// AddUserIdProvider
+// @Title AddUserIdProvider
+// @Tag User API
+// @Description add user id provider
+// @Param   body    body   object.UserIdProvider  true        "The details of the user id provider"
+// @Success 200 {object} controllers.Response The Response object
+// @router /add-user-id-provider [post]
+func (c *ApiController) AddUserIdProvider() {
+	var userIdProvider object.UserIdProvider
+	err := json.Unmarshal(c.Ctx.Input.RequestBody, &userIdProvider)
+	if err != nil {
+		c.ResponseBadRequest(err.Error())
+		return
+	}
+
+	if !c.IsGlobalAdmin() {
+		c.ResponseUnauthorized(c.T("auth:Unauthorized operation"))
+		return
+	}
+
+	if userIdProvider.CreatedTime == "" {
+		userIdProvider.CreatedTime = util.GetCurrentTime()
+	}
+
+	if userIdProvider.Owner == "" || ((userIdProvider.ProviderName == "") == (userIdProvider.LdapId == "")) || userIdProvider.UsernameFromIdp == "" {
```

Comment:
`((userIdProvider.ProviderName == "") == (userIdProvider.LdapId == ""))` This expression corresponds to the logical XOR operation. In a specific example, we check that the entry in the table will reflect the origin of the user, which is what it should be. If the record does not specify either the provider name or the LDAP identifier, then the record is meaningless.

`(userIdProvider.ProviderName == "") && (userIdProvider.LdapId == "")` has a different meaning: it means that both the provider name and the LDAP identifier must be filled in. The record can indicate the origin of the user from the provider **OR** from LDAP.

The condition with XOR satisfies the possible outcomes:

- If both values are empty, then the entry is incorrect.
- If one of the values is filled in, then the entry reflects the origin of the user and is correct.
- Both values are filled in – such a case is impossible, since the purpose of the table is lost – to record _WHEN_ **AND** _FROM_ which source the user came.